tests/integration/query/commits/branchables/peer_update_test.go

```diff
@@ -0,0 +1,260 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package branchables
+
+import (
+	"testing"
+
+	"github.com/sourcenetwork/immutable"
+
+	testUtils "github.com/sourcenetwork/defradb/tests/integration"
+)
+
+func TestQueryCommitsBranchables_HandlesConcurrentUpdatesAcrossPeerConnection(t *testing.T) {
+	test := testUtils.TestCase{
+		Actions: []any{
+			testUtils.RandomNetworkingConfig(),
+			testUtils.RandomNetworkingConfig(),
+			testUtils.SchemaUpdate{
+				Schema: `
+					type Users @branchable {
+						name: String
+					}
+				`,
+			},
+			testUtils.CreateDoc{
+				Doc: `{
+					"name":	"John"
+				}`,
+			},
+			testUtils.UpdateDoc{
+				NodeID: immutable.Some(0),
+				Doc: `{
+					"name":	"Fred"
+				}`,
+			},
+			testUtils.UpdateDoc{
+				NodeID: immutable.Some(1),
+				Doc: `{
+					"name":	"Shahzad"
+				}`,
+			},
+			testUtils.ConnectPeers{
+				SourceNodeID: 1,
+				TargetNodeID: 0,
+			},
+			testUtils.WaitForSync{},
+			testUtils.UpdateDoc{
+				// Update node 1 after the peer connection has been established, this will cause the `Shahzad` commit
+				// to be synced to node 0, as well as the related collection commits.
+				NodeID: immutable.Some(1),
+				Doc: `{
+					"name":	"Chris"
+				}`,
+			},
+			testUtils.WaitForSync{},
+			testUtils.UpdateDoc{
+				// Update node 0 after `Chris` and `Shahzad` have synced to node 0.  As this update happens after the peer
+				// connection has been established, this will cause the `Fred` and `Addo` doc commits, and their corresponding
+				// collection-level commits to sync to node 1.
+				//
+				// Now, all nodes should have a full history, including the 'offline' changes made before establishing the
+				// peer connection.
+				NodeID: immutable.Some(0),
+				Doc: `{
+				  "name": "Addo"
+				}`,
+			},
+			testUtils.WaitForSync{},
+			testUtils.Request{
+				// Strong eventual consistency must now have been established across both nodes, the result of this query
+				// *must* exactly match across both nodes.
+				Request: `query {
+						commits {
+							cid
+							links {
+								cid
+							}
+						}
+					}`,
+				Results: map[string]any{
+					"commits": []map[string]any{
+						{
+							"cid": testUtils.NewUniqueCid("collection, update3"),
+							"links": []map[string]any{
+								{
+									"cid": testUtils.NewUniqueCid("collection, update2"),
```

Comment:
suggestion: If you keep the node in the string (`collection, node0 update2`) it would help to know where this update comes from.

Reply:
Oh true, I dropped that forgetting that the update was actually applied to a single node (at first).  Will do, thanks Fred.

- [ ] note update node in cid description